src/Data/Nat/Primality.agda

```diff
@@ -8,80 +8,331 @@
+  using (_≡_; _≢_; refl; cong)
 
 private
   variable
-    n : ℕ
+    d k m n p : ℕ
+
+  instance
+    nt[2] : NonTrivial 2
+    nt[2] = nonTrivial {0}
 
 ------------------------------------------------------------------------
 -- Definitions
 
+-- Definition of having a non-trivial divisor below a given bound
+
+record HasBoundedNonTrivialDivisor (k n : ℕ) : Set where
+  constructor hasBoundedNonTrivialDivisor
+  field
+    {divisor} : ℕ
+    .{{nt}} : NonTrivial divisor
+    d<k     : divisor < k
+    d∣n     : divisor ∣ n
+
+-- smart constructors
+
+hasBoundedNonTrivialDivisor≢ : .{{NonTrivial d}} → .{{NonZero n}} →
+                    d ≢ n → d ∣ n → HasBoundedNonTrivialDivisor n n
+hasBoundedNonTrivialDivisor≢ d≢n d∣n = hasBoundedNonTrivialDivisor (≤∧≢⇒< (∣⇒≤ d∣n) d≢n) d∣n
+
+hasBoundedNonTrivialDivisor>1 : 1 < d → d < n → d ∣ n → HasBoundedNonTrivialDivisor n n
+hasBoundedNonTrivialDivisor>1 1<d = hasBoundedNonTrivialDivisor
+  where instance _ = n>1⇒nonTrivial 1<d
+
+hasBoundedNonTrivialDivisor∣ : HasBoundedNonTrivialDivisor k m → m ∣ n →
+                               HasBoundedNonTrivialDivisor k n
+hasBoundedNonTrivialDivisor∣ (hasBoundedNonTrivialDivisor d<k d∣m) m∣n
+  = hasBoundedNonTrivialDivisor d<k (∣-trans d∣m m∣n)
+
 -- Definition of compositeness
 
-Composite : ℕ → Set
-Composite 0 = ⊥
-Composite 1 = ⊥
-Composite n = ∃ λ d → 2 ≤ d × d < n × d ∣ n
+Composite : Pred ℕ _
+Composite n = HasBoundedNonTrivialDivisor n n
+
+-- bounded equivalent
+CompositeUpTo : Pred ℕ _
+CompositeUpTo n = ∃⟨ (_< n) ∩ HasNonTrivialDivisor ⟩
+  where
+  HasNonTrivialDivisor : Pred ℕ _
+  HasNonTrivialDivisor d = NonTrivial d × d ∣ n
+
+-- smart constructors
+
+composite : .{{NonTrivial d}} → d < n → d ∣ n → Composite n
+composite {d = d} = hasBoundedNonTrivialDivisor {divisor = d}
+
+composite≢ : ∀ d → .{{NonTrivial d}} → .{{NonZero n}} → d ≢ n → d ∣ n → Composite n
+composite≢ d d≢n d∣n = hasBoundedNonTrivialDivisor≢ {d} d≢n d∣n
+
+composite∣ : .{{NonZero n}} → Composite m → m ∣ n → Composite n
+composite∣ (hasBoundedNonTrivialDivisor {d} d<k d∣n) m∣n@(divides-refl q)
+  = hasBoundedNonTrivialDivisor (*-monoʳ-< q d<k) (*-monoʳ-∣ q d∣n)
+  where instance
+    _ = m≢0∧n>1⇒m*n>1 q d
+    _ = m*n≢0⇒m≢0 q
+
+-- Definition of 'rough': a number is k-rough
+-- if all its non-trivial factors d are bounded below by k
+
+Rough : ℕ → Pred ℕ _
+Rough k n = ¬ HasBoundedNonTrivialDivisor k n
+
+-- Definition of primality: complement of Composite
+-- Constructor `prime` takes a proof isPrime that
+-- NonTrivial p is p-Rough, and thereby enforces that
+-- p is a fortiori NonZero and NonUnit
+
+record Prime (p : ℕ) : Set where
+  constructor prime
+  field
+    .{{nt}}  : NonTrivial p
+    isPrime : Rough p p
+
+-- bounded equivalent
+PrimeUpTo : Pred ℕ _
+PrimeUpTo n = ∀[ (_< n) ⇒ HasNoNonTrivialDivisor ]
+  where
+  HasNoNonTrivialDivisor : Pred ℕ _
+  HasNoNonTrivialDivisor d = NonTrivial d → d ∤ n
+
+-- Definition of irreducibility
+
+module _ (n : ℕ) where
```

Comment:
I think the interoperability cost of adding a module here isn't really worth it? The two main ones should be `Pred` after all, and the private definition should be a `Rel`? To write otherwise is a little confusing!

Reply:
OK! (But further evidence of how my strange design-mind operates... so good to have that checked/litmus-tested)